File: backends/reference/kernels/ffn_gated.py
"""ffn.gated@1.0.0 — act(x·gateᵀ) ⊙ (x·upᵀ) · outᵀ, with optional biases.

| branch / record        | status                       |
|------------------------|------------------------------|
| activation silu        | implemented                  |
| activation gelu        | implemented (erf)            |
| activation gelu_tanh   | implemented                  |
| activation relu2       | implemented                  |
| in_bias, out_bias      | implemented                  |
| condition_dim / condition | refused (M2+)             |
| activation_sparsity    | refused when > 0             |
"""
import torch
import torch.nn.functional as F
from kernels._common import present, refuse_unknown, w

CONTRACT = ("ffn.gated", "1.0.0")
KNOWN = {'width', 'inner', 'activation', 'in_bias', 'out_bias', 'condition_dim', 'condition', 'activation_sparsity'}
ACT = {'silu': F.silu, 'gelu': F.gelu, 'gelu_tanh': lambda x: F.gelu(x, approximate='tanh'),
       'relu2': lambda x: F.relu(x).pow(2)}


def supports(arguments):
    reasons = []
    refuse_unknown(arguments, KNOWN, reasons)
    if arguments.get('activation') not in ACT:
        reasons.append(f"activation={arguments.get('activation')}")
    if present(arguments, 'condition_dim'):
        reasons.append(f"condition_dim={arguments['condition_dim']}")
    if present(arguments, 'condition'):
        reasons.append(f"condition={arguments['condition']}")
    if arguments.get('activation_sparsity'):
        reasons.append(f"activation_sparsity={arguments['activation_sparsity']}")
    return reasons


def run(ctx, arguments, inputs, params, states):
    x = inputs['input']
    g = x @ w(ctx, params['gate']).T
    u = x @ w(ctx, params['up']).T
    if arguments.get('in_bias'):
        g = g + w(ctx, params['gate_bias'])
        u = u + w(ctx, params['up_bias'])
    h = ACT[arguments['activation']](g) * u
    y = h @ w(ctx, params['out']).T
    if arguments.get('out_bias'):
        y = y + w(ctx, params['out_bias'])
    return {'output': y}

File: generators/zml/primitives/ffn_gated.zig
//! ffn.gated@1.0.0 — act(x·gateᵀ) ⊙ (x·upᵀ) · outᵀ, with optional biases.
//!
//! | branch                   | status                                  |
//! |--------------------------|-----------------------------------------|
//! | activation silu          | implemented                             |
//! | activation gelu_tanh     | implemented                             |
//! | activation relu2         | implemented                             |
//! | activation gelu (erf)    | refused — ZML's `gelu` is the tanh approximation, which the contract names `gelu_tanh` and distinguishes from the exact one; mapping by name would compute the wrong function silently |
//! | in_bias, out_bias        | implemented                             |
//! | activation_sparsity      | refused when > 0                        |

const std = @import("std");

const zml = @import("zml");

const p = @import("../primitive.zig");

pub const primitive: p.Primitive = .{
    .name = "ffn.gated",
    .version = "1.0.0",
    .run = run,
    .capabilities =
    \\{"arguments": {"width": "any", "inner": "any",
    \\               "activation": ["silu", "gelu_tanh", "relu2"],
    \\               "in_bias": "any", "out_bias": "any",
    \\               "activation_sparsity": {"absent": true, "values": [0, 0.0]}},
    \\ "states": [],
    \\ "notes": ["activation gelu (erf) is refused: ZML's gelu is the tanh approximation"]}
    ,
};

fn activate(name: []const u8, x: zml.Tensor) !zml.Tensor {
    if (std.mem.eql(u8, name, "silu")) return x.silu();
    if (std.mem.eql(u8, name, "gelu_tanh")) return x.gelu();
    if (std.mem.eql(u8, name, "relu2")) return x.relu().powByConst(2);
    return p.Error.Unimplemented;
}

fn run(ctx: *p.Ctx, call: p.Call) ![]const p.Binding {
    const x = p.features(call.inputs.must("input"));
    const activation = call.argStr("activation") orelse return p.Error.MissingArgument;

    var g = p.linear(x, call.params.must("gate"));
    var u = p.linear(x, call.params.must("up"));
    if (call.argBool("in_bias")) {
        g = g.add(p.features(call.params.must("gate_bias")).convert(g.dtype()).broad(g.shape()));
        u = u.add(p.features(call.params.must("up_bias")).convert(u.dtype()).broad(u.shape()));
    }

    var y = p.linear((try activate(activation, g)).mul(u), call.params.must("out"));
    if (call.argBool("out_bias")) {
        y = y.add(p.features(call.params.must("out_bias")).convert(y.dtype()).broad(y.shape()));
    }
    return p.one(ctx, "output", y);
}
